src/vulkan/engine_vulkan_debugger.cc

```diff
@@ -0,0 +1,976 @@
+      return;
+    }
+
+    amber::debug::Location got_location;
+    std::string got_source_line;
+    if (!client_.FrameLocation(frame, &got_location, &got_source_line)) {
+      return;
+    }
+
+    if (got_location.file != location.file) {
+      OnError("Expected file to be '" + location.file + "' but file was " +
+              got_location.file);
+    } else if (got_location.line != location.line) {
+      OnError("Expected line number to be " + std::to_string(location.line) +
+              " but line number was " + std::to_string(got_location.line));
+    } else if (line != "" && got_source_line != line) {
+      OnError("Expected source line to be:\n  " + line + "\nbut line was:\n  " +
+              got_source_line);
+    }
+  }
+
+  void ExpectLocal(const std::string& name, int64_t value) override {
+    DEBUGGER_LOG("ExpectLocal('%s', %d)", name.c_str(), (int)value);
+    ExpectLocalT(name, value);
+  }
+
+  void ExpectLocal(const std::string& name, double value) override {
+    DEBUGGER_LOG("ExpectLocal('%s', %f)", name.c_str(), value);
+    ExpectLocalT(name, value);
+  }
+
+  void ExpectLocal(const std::string& name, const std::string& value) override {
+    DEBUGGER_LOG("ExpectLocal('%s', '%s')", name.c_str(), value.c_str());
+    ExpectLocalT(name, value);
+  }
+
+  template <typename T>
+  void ExpectLocalT(const std::string& name, const T& expect) {
+    dap::StackFrame frame;
+    if (!client_.TopStackFrame(threadId_, &frame)) {
+      return;
+    }
+
+    Variables locals;
+    if (!client_.GetLocals(frame, &locals)) {
+      return;
+    }
+
+    if (auto lane = client_.GetLane(locals, lane_)) {
+      auto owner = lane;
+      const Variable* var = nullptr;
+      std::string path;
+      for (auto part : Split(name, ".")) {
+        var = owner->Find(part);
+        if (!var) {
+          if (owner == lane) {
+            OnError("Local '" + name + "' not found\nAll Locals: " +
+                    lane->AllNames() + ".\nLanes: " + locals.AllNames() + ".");
+          } else {
+            OnError("Local '" + path + "' does not contain '" + part +
+                    "'\nChildren: " + owner->AllNames());
+          }
+          return;
+        }
+        owner = &var->children;
+        path += (path.size() > 0) ? "." + part : part;
+      }
+
+      T got = {};
+      if (!var->Get(&got)) {
+        OnError("Local '" + name + "' was not of expected type");
+        return;
+      }
+
+      if (got != expect) {
+        std::stringstream ss;
+        ss << "Local '" << name << "' did not have expected value. Value is '"
+           << got << "', expected '" << expect << "'";
+        OnError(ss.str());
+        return;
+      }
+    }
+  }
+
+ private:
+  void OnError(const std::string& err) {
+    DEBUGGER_LOG("ERROR: %s", err.c_str());
+    error_ += err;
+  }
+
+  const dap::integer threadId_;
+  const int lane_;
+  Client client_;
+  std::thread thread_;
+  Event done_;
+  Result error_;
+};
+
+}  // namespace
```

Comment:
Move down to 664 so anonymous namespace is inside amber vulkan namespace.

Reply:
Done (I think this is what you wanted)

Reply:
yup